### PR TITLE
Add separate-space dithering

### LINKED ISSUE #23: Add per-channel dithering
Add a meta-method that takes any gray-scale dithering algorithm and applies channel-wise binary dithering to color images using `channelview`.

An illustration of this is shown [here](http://caca.zoy.org/wiki/libcaca/study/6), where it is called "separate-space dithering".

The type `AbstractColorDither` will need be extended by `AbstractCustomColorDither` to differentiate this method from other methods like Floyd-Steinberg that take custom user-defined color palettes.

The output image will look different depending on the concrete color type on which `channelview` is applied. As RGB is typically used, this should be the default.

**Pseudocode example:**
```julia
alg = SeparateSpaceDither(Bayer(); colorspace=RGB)
dither!(img, alg)
```

### PR DESCRIPTION
**Pull request by adrhill**
Implements the `SeparateSpace` method including tests and examples in the docs. Closes #23.

## API changes
This PR necessitated a small reworking of the API. The type hierarchy now looks as follows:
```julia
# Algorithms for which the user can define a custom output color palette.
# If no palette is specified, algorithms of this type will default to a binary color palette
# and act similarly to algorithms of type AbstractBinaryDither.
abstract type AbstractCustomColorDither <: AbstractDither end
```
```julia
# Algorithms whose output uses predefined sets of color, e.g. separate space dithering:
abstract type AbstractFixedColorDither <: AbstractDither end
```
```julia
# Algorithms which strictly do binary dithering:
abstract type AbstractBinaryDither <: AbstractFixedColorDither end
```

As the docstrings point out, every `AbstractCustomColorDither` algorithm is expected to work like a `AbstractBinaryDither` algorithm when applied to a gray-scale image without providing a color scheme. This is necessary to be able to use error diffusion algorithms (which are of type `AbstractCustomColorDither`) with `SeparateSpace`.

## Breaking changes
The default output type of `dither` on gray-scale images is now the image `eltype` instead of `Gray{Bool}`.

### CODE REVIEW

File: src/separate_space.jl [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+A meta-algorithm that takes any gray-scale dithering algorithm and applies channel-wise
+binary dithering.
+
+# Note:
+The output of this algorithm depends on the color type of input image. RGB is recommended.
+"""
+struct SeparateSpace{A<:AbstractDither} <: AbstractFixedColorDither
+    alg::A
+end
```
Comment (johnnychen94) [inline src/separate_space.jl:12]: It would be nice to have a piece of small example here, maybe?

**Review by johnnychen94 — APPROVED** [review]
Just curious, will

```julia
alg = Bayer() |> SeparateSpace |> SeparateSpace |> SeparateSpace |> SeparateSpace
```

work? 😄

**Comment by adrhill** [comment]
> Just curious, will
> 
> ```julia
> alg = Bayer() |> SeparateSpace |> SeparateSpace |> SeparateSpace |> SeparateSpace
> ```
> 
> work? 😄

I expected this to be the same as `Bayer() |> SeparateSpace`, but it isn't:
image
This is quite strange, as `SeparateSpace` just uses views onto `channelview`. At some point the images passed through `SeparateSpace` should just have one channel and nothing should change, but this doesn't appear to be the case.

This gets worse with repeated applications:
image

**Comment by adrhill** [comment]
The issue is that repeated application of `view(img, 1, :, :)` slices the image:
```julia-repl
julia> img = rand(RGB, 5, 5)
5×5 Array{RGB{Float64},2} with eltype RGB{Float64}:
 RGB{Float64}(0.715113,0.491371,0.296905)   …  RGB{Float64}(0.791234,0.951492,0.814854)
 RGB{Float64}(0.72507,0.188623,0.939496)       RGB{Float64}(0.743093,0.680352,0.643163)
 RGB{Float64}(0.706082,0.0881292,0.518988)     RGB{Float64}(0.158277,0.374302,0.0810188)
 RGB{Float64}(0.731997,0.700499,0.676473)      RGB{Float64}(0.230681,0.79649,0.11652)
 RGB{Float64}(0.780349,0.169881,0.798904)      RGB{Float64}(0.659169,0.948437,0.774215)

julia> channelview(view(channelview(img), 1, :, :))
5×5 view(reinterpret(reshape, Float64, ::Array{RGB{Float64},2}), 1, :, :) with eltype Float64:
 0.715113  0.134762  0.637772   0.213     0.791234
 0.72507   0.211181  0.925992   0.289092  0.743093
 0.706082  0.356741  0.0447958  0.523512  0.158277
 0.731997  0.747477  0.848042   0.737581  0.230681
 0.780349  0.916575  0.468991   0.709834  0.659169

julia> channelview(view(channelview(view(channelview(img), 1, :, :)), 1, :, :))
5×1 view(reshape(view(reinterpret(reshape, Float64, ::Array{RGB{Float64},2}), 1, :, :), 5, 5, 1), 1, :, :) with eltype Float64:
 0.71511314161005
 0.13476170768648266
 0.6377716535250101
 0.21300024258756445
 0.791234444784436
 ```

**Comment by johnnychen94** [comment]
How about adding a constructor to avoid such cases?

```julia
SeparateSpace(alg::SeparateSpace) = alg
```

**Comment by adrhill** [comment]
That's a more elegant solution than I had in mind... 😄

**Pull request merged**